esp-hal/src/work_queue.rs

```diff
@@ -0,0 +1,459 @@
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Status {
+    /// The processing has completed.
+    Completed,
+
+    /// The work item has been cancelled.
+    Cancelled,
+}
+
+/// A unit of work in the work queue.
+pub(crate) struct WorkItem<T: Sync> {
+    next: Option<NonNull<WorkItem<T>>>,
+    status: Poll,
+    data: T,
+    waker: AtomicWaker,
+}
+
+impl<T: Sync> WorkItem<T> {
+    /// Completes the work item.
+    ///
+    /// This function is intended to be called from the underlying drivers.
+    pub fn complete(&mut self, status: Status) {
+        self.status = Poll::Ready(status);
+        self.waker.wake();
+    }
+
+    /// Prepares a work item to be enqueued.
+    ///
+    /// # Safety:
+    ///
+    /// The caller must ensure the reference is not used again while the pointer returned by this
+    /// function is in use.
+    unsafe fn prepare(&mut self) -> NonNull<Self> {
+        self.next = None;
+        self.status = Poll::Pending;
+
+        NonNull::from(self)
+    }
+}
+
+/// The status of a work item posted to a work queue.
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Poll {
+    /// The work item has not yet been fully processed.
+    Pending,
+
+    /// The work item has been processed.
+    Ready(Status),
+}
+
+/// A reference to a posted [`WorkItem`].
+///
+/// This struct ensures that the work item is valid until the item is processed or is removed from
+/// the work queue.
+///
+/// Dropping the handle cancels the work item, but may block for some time if the work item is
+/// already being processed.
+pub(crate) struct Handle<'t, T: Sync> {
+    queue: &'t WorkQueue<T>,
+    work_item: NonNull<WorkItem<T>>,
+    // Make sure lifetime is invariant to prevent UB.
+    _marker: PhantomData<&'t mut WorkItem<T>>,
+}
+
+impl<'t, T: Sync> Handle<'t, T> {
+    fn status(&self) -> Poll {
+        unsafe { (*self.work_item.as_ptr()).status }
+    }
+
+    /// Returns the status of the work item.
+    pub fn poll(&mut self) -> Poll {
+        let status = self.status();
+        if status == Poll::Pending {
+            self.queue.process();
+            self.status()
+        } else {
+            status
+        }
+    }
+
+    /// Waits for the work item to be processed.
+    pub fn wait(&mut self) -> impl Future<Output = Status> {
+        poll_fn(|ctx| {
+            unsafe { (*self.work_item.as_ptr()).waker.register(ctx.waker()) };
+            match self.poll() {
+                Poll::Pending => core::task::Poll::Pending,
+                Poll::Ready(status) => core::task::Poll::Ready(status),
+            }
+        })
+    }
+}
+
+impl<'t, T: Sync> Drop for Handle<'t, T> {
+    fn drop(&mut self) {
+        if self.status() == Poll::Pending {
+            self.queue.cancel(self.work_item);
+            // We must wait for the driver to release our WorkItem.
+            while self.poll() == Poll::Pending {}
```

Comment:
Is it possible for `Handle` to be dropped in a high priority task whilst a low priority task is processing the item/cancellation, leading to deadlock?

Reply:
I don't think it should ever be possible to deadlock this. The point of this design is that any poller will kick the queue forward. Obviously this will need to be tested, though.